docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,232 @@
+  typical Linen train step.
+
+
+# Previous API
+[previous API]: #previous-api
+
+An optimizer would first be constructed from its definition and the pytree of
+target params:
+
+```python
+optimizer_def = flax.optim.Momentum(learning_rate=0.1, beta=0.9)
+optimizer = optimizer_def.create(variables['params'])
+```
+
+The target variables would then optimized in the train step (assuming a single
+non-params collection "batch_stats"):
+
+```python
+def make_train_step(apply_fn):
+  @jax.jit
+  def train_step(optimizer, batch_stats, inputs, labels):
+
+    def loss_fn(params):
+      variables = {'params': params, 'batch_stats': batch_stats}
+      logits, new_model_state = apply_fn(
+          variables, inputs, mutable=['batch_stats'])
+      loss = xent_loss(logits, labels)
+      return loss, new_model_state['batch_stats']
+
+    (loss, new_batch_stats), grad = jax.value_and_grad(loss_fn, has_aux=True)(
+        optimizer.target)
+    lr = get_learning_rate(step)
+    new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+    return new_optimizer, new_batch_stats, loss
+
+  return train_step
+
+
+batch_stats = variables['batch_stats']
+train_step = make_train_step(model.apply)
+for step, batch in enumerate(ds)
+  optimizer, batch_stats, loss = train_step(
+      optimizer, batch_stats, batch['image'], batch['label'])
+```
+
+The optimizer itself would be implemented by creating a new class deriving
+from `OpimizerDef`:
+
+```python
+# flax/optim/momentum.py
+
+@flax.struct.dataclass
+class _MomentumHyperParams:
+  learning_rate: np.ndarray
+  beta: np.ndarray
+
+
+@flax.struct.dataclass
+class _MomentumParamState:
+  momentum: onp.ndarray
+
+
+class Momentum(flax.optim.OptimizerDef):
+
+  def __init__(self, learning_rate=None, beta=0.9):
+    super().__init__(
+      _MomentumHyperParams(learning_rate, beta)
+    )
+
+  def init_param_state(self, param):
+    return _MomentumParamState(jnp.zeros_like(param))
+
+  def apply_param_gradient(self, step, hyper_params, param, state, grad):
+    del step
+    assert hyper_params.learning_rate is not None
+    new_momentum = state.momentum * hyper_params.beta + grad
+    new_params = param - hyper_params.learning_rate * new_momentum
+    return new_params, _MomentumParamState(new_momentum)
+```
+
+Remarks:
+
+- The functions `init_param_state()` and `apply_param_gradient()` are called
+  for every leaf in the params/grads pytree. This makes it possible to write the
+  calculations directly without `jax.tree_multimap()`.
+- The interface was defined in pre-Linen without the distinction of `params` vs.
+  other collections in `variables` in mind. The original API was elegant because
+  one only needed to pass around the optimizer, which included the parameters,
+  optimizer state, optimizer hyperparameters, and a reference to the
+  `OptimizerDef` to perform the param/state update.
+- Note that the current implementation contains some additional functionality
```

Comment:
This is another potential gap with Optax (at least at the moment) -- if we consider switching to Optax, we should find what the new pattern for `MultiOptimizer` and passing in `focus`, and whether we can (or should) keep a backwards compatibility layer.

Reply:
Refactored the entire FLIP. Now Optax is in motivation & has its own subsection. See also #1053.